llvm/lib/CheerpUtils/PartialExecuter.cpp

```diff
@@ -0,0 +1,1777 @@
+		}
+		i++;
+		}
+
+			}
+		}
+//		if (isa<PtrToIntInst>(I))
+//			return true;	//PTR TO INT ARE not preexecutable
+//TODO: also int to ptr ??
+		if (StoreInst* load = dyn_cast<StoreInst>(&I))
+		{
+			return true;
+  GenericValue SRC = getOperandValue(load->getPointerOperand(), getLastStack());
+  GenericValue *Ptr = (GenericValue*)GVTORP(SRC);
+if ((long long)Ptr < 100)
+	return true;
+		}
+		if (LoadInst* load = dyn_cast<LoadInst>(&I))
+		{
+		//	return true;
+  GenericValue SRC = getOperandValue(load->getPointerOperand(), getLastStack());
+  GenericValue *Ptr = (GenericValue*)GVTORP(SRC);
+		 
+std::vector<std::pair<long long, long long> > V;
+for (auto& gv : I.getFunction()->getParent()->globals())
+{
+	GlobalVariable* GV = dyn_cast<GlobalVariable>(&gv);
+	if (!GV)
+		continue;
+	if (!GV->hasInitializer())
+		continue;
+	if (GV->isExternallyInitialized())
+		continue;
+  GenericValue SRC = getOperandValue(GV, getLastStack());
+  GenericValue *Ptr = (GenericValue*)GVTORP(SRC);
+//llvm ::errs() << GV->getName() << "\t" <<"\t" << (long long)Ptr<<"\t"<< GV->isConstant()<<"\n";
+	//bail out if there are pointers 
+//	llvm::errs() << *GV->getInitializer() << "\n";
+if (GV->isConstant())
+  V.push_back({(long long)Ptr, ((long long)Ptr)+getDataLayout().getTypeAllocSize(GV->getInitializer()->getType())});
+}
+
+//std::sort(V.begin(), V.end());
+for (int i=0; i<V.size(); i++)
+{
+	if ((long long)Ptr >= V[i].first && (long long)Ptr < V[i].second)
+		return false;
+}
+
+if (false)
+if (load->getPointerOperand()->getName() == "flags.i" ||
+		load->getPointerOperand()->getName() == "width.i" ||
+		load->getPointerOperand()->getName() == "prec.i" ||
+		load->getPointerOperand()->getName() == "dprec.i" ||
+		load->getPointerOperand()->getName() == "lead.i")
+
+{
+//	llvm::errs() << I << "\n";
+	return false;
+}
+
+
+return true;
+
+
+
+			Value* ptr = load->getPointerOperand();
+			if (GetElementPtrInst* gep = dyn_cast<GetElementPtrInst>(ptr))
+			{
+				if (!gep->isInBounds())
+					return true;
+				//TODO: CHECK whether the location is also constant!!
+				return false;
+			}
+			//ConstantExpr are also possibly good
+			//PHI also
+			//
+			//In reality here we need a map of valid ptrs to be updated
+			//TODO: 
+			return false;
+		}	
+		return false;
+	}
+void clearFunction(llvm::Function& F)
+{
+	for (Instruction& I : instructions(F))
+	{
+		visitOuter<VisitingPolicy::REMOVE_VALUES>(I);
+	}
+}
+
+template <PartialInterpreter::VisitingPolicy policy>
+void visitOuter(llvm::Instruction& I) 
+	{
+			if (policy == REMOVE_VALUES)
+			{
+				if (computed.count(&I))
+					computed.erase(&I);
+				return;
+			}
+		
+		if (PHINode* phi = dyn_cast<PHINode>(&I))
+		{
+			const llvm::BasicBlock* from = getIncomingBB();
+			if (from)
+			{
+				llvm::Value* incoming = phi->getIncomingValueForBlock(from);
+				assert(incoming);
+				if (isValueComputed(incoming))
+					incomings.push_back({phi, {getOperandValue(incoming, getLastStack()), getBitMask(incoming)}});
```

Comment:
This seems to be just used to populate the computed map at soon as the next inst is executed. Can we just populate it right away and avoid the whole "incoming" member